### PR TITLE
Controllers Branch - End to end functionality

### PR DESCRIPTION
**Pull request by jherico**
This PR manages to let you navigate (using an SDL2 supported controller) entirely via the new code.  

It contains a lot of commented out code from the old version that needs to be ported to JSON files, which I'm in the process of doing.

### CODE REVIEW

File: interface/src/devices/3DConnexionClient.cpp [hunk 1]
```diff
@@ -48,76 +52,71 @@ void ConnexionData::setButton(int lastButtonState) {
...
-    // Zoom
-    //  mapper.addInputChannel(UserInputMapper::BOOM_IN, makeInput(ROTATION_AXIS_Z_NEG), BOOM_SPEED);
-    //  mapper.addInputChannel(UserInputMapper::BOOM_OUT, makeInput(ROTATION_AXIS_Z_POS), BOOM_SPEED);
-
+QString ConnexionData::getDefaultMappingConfig() {
+    static const QString MAPPING_JSON = PathUtils::resourcesPath() + "/controllers/vive.json";
+    return MAPPING_JSON;
 }
 
+//void ConnexionData::assignDefaultInputMapping(UserInputMapper& mapper) {
```
Comment (jherico) [inline interface/src/devices/3DConnexionClient.cpp]: Pending migration to JSON

File: libraries/controllers/src/controllers/ScriptingInterface.cpp [hunk 2]
```diff
@@ -26,446 +24,186 @@
...
-        return StandardButtonChannel::NUM_STANDARD_BUTTONS;
-    }
-
-    bool ScriptingInterface::isButtonPressed(int buttonIndex) const {
-        return getButtonValue((StandardButtonChannel)buttonIndex) == 0.0f ? false : true;
-    }
-
-    int ScriptingInterface::getNumberOfTriggers() const {
-        return StandardCounts::TRIGGERS;
+    //bool ScriptingInterface::isPrimaryButtonPressed() const {
```
Comment (jherico) [inline libraries/controllers/src/controllers/ScriptingInterface.cpp]: Are these API's needed?  We need to port them in our JS if they're in use.  We may need to retain for a bit if they're used by lots of user scripts.

File: libraries/controllers/src/controllers/StandardController.cpp [hunk 3]
```diff
@@ -26,143 +35,111 @@ void StandardController::focusOutEvent() {
...
-    return UserInputMapper::Input(_deviceID, button, UserInputMapper::ChannelType::BUTTON);
+QString StandardController::getDefaultMappingConfig() {
+    static const QString DEFAULT_MAPPING_JSON = PathUtils::resourcesPath() + "/controllers/standard.json";
+    return DEFAULT_MAPPING_JSON;
 }
 
-UserInputMapper::Input StandardController::makeInput(controller::StandardAxisChannel axis) {
-    return UserInputMapper::Input(_deviceID, axis, UserInputMapper::ChannelType::AXIS);
-}
+// FIXME figure out how to move the shifted version to JSON
```
Comment (jherico) [inline libraries/controllers/src/controllers/StandardController.cpp]: Pending migration to JSON...   actually I think this one may be done, but we don't support the 'shoulder button for precision movement' functionality, so all these have modifiers.

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 4]
```diff
@@ -402,56 +403,18 @@ void UserInputMapper::assignDefaulActionScales() {
     _actionScales[YAW] = 1.0f; // default
 }
 
-// This is only necessary as long as the actions are hardcoded
-// Eventually you can just add the string when you add the action
-void UserInputMapper::createActionNames() {
```
Comment (jherico) [inline libraries/controllers/src/controllers/UserInputMapper.cpp]: Handled by the new ActionDevice

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 5]
```diff
@@ -475,38 +438,435 @@ QScriptValue actionToScriptValue(QScriptEngine* engine, const UserInputMapper::A
...
 }
 
-UserInputMapper::Input UserInputMapper::makeStandardInput(controller::StandardPoseChannel pose) {
+Input UserInputMapper::makeStandardInput(controller::StandardPoseChannel pose) {
     return Input(STANDARD_DEVICE, pose, ChannelType::POSE);
 }
+
+
+
+void UserInputMapper::update() {
```
Comment (jherico) [inline libraries/controllers/src/controllers/UserInputMapper.cpp]: moved from the scripting interface

**Comment by hifi-gustavo** [comment]
Build console output for commit: 14f26e8a29dc7dd4ec4264000a9aa1d8e3b11a43 available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

File: libraries/input-plugins/src/input-plugins/KeyboardMouseDevice.cpp [hunk 6]
```diff
@@ -128,163 +130,159 @@ void KeyboardMouseDevice::touchUpdateEvent(const QTouchEvent* event) {
...
-    
-    mapper.addInputChannel(UserInputMapper::SHIFT, makeInput(Qt::Key_Space));
-    mapper.addInputChannel(UserInputMapper::ACTION1, makeInput(Qt::Key_R));
-    mapper.addInputChannel(UserInputMapper::ACTION2, makeInput(Qt::Key_T));
+QString KeyboardMouseDevice::getDefaultMappingConfig() {
+    static const QString MAPPING_JSON = PathUtils::resourcesPath() + "/controllers/vive.json";
+    return MAPPING_JSON;
 }
 
+//void KeyboardMouseDevice::assignDefaultInputMapping(UserInputMapper& mapper) {
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/KeyboardMouseDevice.cpp]: pending migration to JSON

File: libraries/input-plugins/src/input-plugins/SixenseManager.cpp [hunk 7]
```diff
@@ -456,85 +458,91 @@ static const auto R2 = controller::A;
...
-    mapper.addInputChannel(UserInputMapper::CONTEXT_MENU, makeInput(L0));
-    mapper.addInputChannel(UserInputMapper::TOGGLE_MUTE, makeInput(R0));
 
+QString SixenseManager::getDefaultMappingConfig() {
+    static const QString MAPPING_JSON = PathUtils::resourcesPath() + "/controllers/hydra.json";
+    return MAPPING_JSON;
 }
 
+//
+//void SixenseManager::assignDefaultInputMapping(UserInputMapper& mapper) {
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/SixenseManager.cpp]: pending migration to JSON

File: libraries/input-plugins/src/input-plugins/ViveControllerManager.cpp [hunk 8]
```diff
@@ -386,92 +385,93 @@ void ViveControllerManager::handlePoseEvent(const mat4& mat, int index) {
...
+            makePair(RT, "RT"),
+
+            makePair(LB, "LB"),
+            makePair(RB, "RB"),
+
+            makePair(LS, "LS"),
+            makePair(RS, "RS"),
+        };
+
+        //availableInputs.append(Input::NamedPair(makeInput(LEFT_HAND), "Left Hand"));
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/ViveControllerManager.cpp]: Need to come up with proper names for all the Vive controls.

File: libraries/input-plugins/src/input-plugins/ViveControllerManager.cpp [hunk 9]
```diff
@@ -386,92 +385,93 @@ void ViveControllerManager::handlePoseEvent(const mat4& mat, int index) {
...
-    return UserInputMapper::Input(_deviceID, axis | (index == 0 ? LEFT_MASK : RIGHT_MASK), UserInputMapper::ChannelType::AXIS);
+QString ViveControllerManager::getDefaultMappingConfig() {
+    static const QString MAPPING_JSON = PathUtils::resourcesPath() + "/controllers/vive.json";
+    return MAPPING_JSON;
 }
 
-UserInputMapper::Input ViveControllerManager::makeInput(JointChannel joint) {
-    return UserInputMapper::Input(_deviceID, joint, UserInputMapper::ChannelType::POSE);
-}
+//void ViveControllerManager::assignDefaultInputMapping(UserInputMapper& mapper) {
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/ViveControllerManager.cpp]: pending migration to JSON

**Comment by hifi-gustavo** [comment]
Build console output for commit: b43873a17770c905eb06450c06892821cf878e61 available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bab5a5b25dc5267e260eb3c55190036293b738b9 available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Comment by ZappoMan** [comment]
```
/Users/jenkins/Jenkins/workspace/pull-request-osx/interface/src/devices/3DConnexionClient.cpp:892:27: error: no member named 'registerToUserInputMapper' in 'ConnexionData'
            connexiondata.registerToUserInputMapper(*userInputMapper);
            ~~~~~~~~~~~~~ ^
/Users/jenkins/Jenkins/workspace/pull-request-osx/interface/src/devices/3DConnexionClient.cpp:893:27: error: no member named 'assignDefaultInputMapping' in 'ConnexionData'
            connexiondata.assignDefaultInputMapping(*userInputMapper);
            ~~~~~~~~~~~~~ ^
/Users/jenkins/Jenkins/workspace/pull-request-osx/interface/src/devices/3DConnexionClient.cpp:914:27: error: no member named 'setDeviceID' in 'ConnexionData'
            connexiondata.setDeviceID(0);
            ~~~~~~~~~~~~~ ^
/Users/jenkins/Jenkins/workspace/pull-request-osx/interface/src/devices/3DConnexionClient.cpp:924:23: error: no member named 'registerToUserInputMapper' in 'ConnexionData'
        connexiondata.registerToUserInputMapper(*userInputMapper);
        ~~~~~~~~~~~~~ ^
/Users/jenkins/Jenkins/workspace/pull-request-osx/interface/src/devices/3DConnexionClient.cpp:925:23: error: no member named 'assignDefaultInputMapping' in 'ConnexionData'
        connexiondata.assignDefaultInputMapping(*userInputMapper);
        ~~~~~~~~~~~~~ ^
/Users/jenkins/Jenkins/workspace/pull-request-osx/interface/src/devices/3DConnexionClient.cpp:936:23: error: no member named 'setDeviceID' in 'ConnexionData'
        connexiondata.setDeviceID(0);
        ~~~~~~~~~~~~~ ^
6 errors generated.
```

**Comment by howard-stearns** [comment]
It would be awesome if we came out of this with OS X not crashing on exit in release distributions any more. Thanks!

**Comment by ZappoMan** [comment]
I am about 90% certain that the crash on exit on the mac is a Jenkins/build
system problem. Hopefully Leo has this on his list.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d92a493d8153b7a1d372df2c0d6670cb6b2d49e7 available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a94a6a1f632a5f00061a4de8268f0ebf45155cad available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 10]
```diff
@@ -475,38 +434,361 @@ QScriptValue actionToScriptValue(QScriptEngine* engine, const UserInputMapper::A
...
+
+    EndpointSet readEndpoints;
+    EndpointSet writtenEndpoints;
+
+    static const int HARDWARE_PASS = 0;
+    static const int STANDARD_PASS = 1;
+
+    // Now process the current values for each level of the stack
+    for (auto& mapping : _activeMappings) {
+        for (int pass = 0; pass < 2; ++pass) {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/UserInputMapper.cpp]: magic numbers?
what's 2?
should 0 be HARDWARE_PASS?

**Comment by hifi-gustavo** [comment]
Build console output for commit: ef6c4f6f665c56bd37ae31dede4241a424a95762 available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bea6fdd8902327ad3e1d820d616a4e20d25692f2 available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 91a32a2990ba8f962ed9f0bdb5f072474a94677b available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 22139931f252966396538fb75ef2bd091c89470b available at: https://gist.github.com/1e3a41d4d983752189bd<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6121.dmg

**Pull request merged**